torch/distributed/algorithms/model_averaging/utils.py

```diff
@@ -19,18 +19,45 @@ def average_parameters(
     if dist._rank_not_in_group(group_to_use):
         return
 
+    if isinstance(params, types.GeneratorType):
+        # compatible with model.parameters() input
+        flat_params_all_reduce(params, group_to_use)
+    elif isinstance(params, list):
+        param_groups = params
+        # support optim.param_group input
+        if not comm_memory_efficient:
+            params_list = get_param_list(param_groups)
+            flat_params_all_reduce(params_list, group_to_use)
+        else:
```

Comment:
Let's avoid adding the complexity of such so-called "memory-efficient" implementation. IMO, this will not ship much value, and I don't know anyone will choose it, because:
1) When model averaging like localSGD is applied, the purpose is mainly saving communication cost typically caused by a large model size.
2) Here the "memory-efficient" implementation actually can lead to prohibitively expensive communication cost, because a large number of layers can cause a large number of allreduce calls over probably small tensors. This will contradict with the purpose of model averaging approaches like local SGD.
3) The increase of memory cost here is often acceptable since such flattened tensor is only consumes some extra memory temporarily.

Sorry if I gave you any confusion in my previous comment.

Reply:
agree.   origin code actually provides two methods.
1. comm_memory_efficient:     PostLocalSGDOptimizer style,               for params in param_group["params"]:
                if params.grad is None:
                    continue
                self.averager.average_parameters(iter(params))

Reply:
so your opinion is to delete memory efficient scheme?  
ok, i will fix my cr later.
but maybe some very big model need memory effiecient method in the future. this leaves future.

Reply:
Yeah, my suggestion is completely dropping memory efficient method here, unless you can find a real use case that can benefit from the memory efficient method.

IMO, if we really want to implement a memory efficient method, we should create a fix memory buffer, and flatten tensors into such a buffer for a few times. Therefore, the user can balance the # of all reduce calls and the memory footprint. Right now the implementations are two extremes: doing either a single allreduce or as many allreduces as the # of parameters.

Reply:
+1 on separating out memory efficient argument and implementation into another PR which can possibly be stacked on top of this one for additional discussion and benchmarks about the improvement.

It will also simplify this PR to focus on only the accuracy improvement.

Reply:
https://github.com/pytorch/pytorch/pull/74894#discussion_r842465678